src/Common/tests/System/Net/Http/GenericLoopbackServer.cs

```diff
@@ -0,0 +1,118 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace System.Net.Test.Common
+{
+    // Loopback server abstraction. 
+    // Tests that want to run over both HTTP/1.1 and HTTP/2 should use this instead of the protocol-specific loopback servers.
+
+    public abstract class LoopbackServerFactory
+    {
+        public abstract Task CreateServerAsync(Func<GenericLoopbackServer, Uri, Task> funcAsync);
+
+        public abstract bool IsHttp11 { get; }
+        public abstract bool IsHttp2 { get; }
+
+        // Common helper methods
+
+        public Task CreateClientAndServerAsync(Func<Uri, Task> clientFunc, Func<GenericLoopbackServer, Task> serverFunc)
+        {
+            return CreateServerAsync(async (server, uri) =>
+            {
+                Task clientTask = clientFunc(uri);
+                Task serverTask = serverFunc(server);
+
+                await new Task[] { clientTask, serverTask }.WhenAllOrAnyFailed();
+            });
+        }
+    }
+
+    public sealed class Http11LoopbackServerFactory : LoopbackServerFactory
+    {
+        public static readonly Http11LoopbackServerFactory Singleton = new Http11LoopbackServerFactory();
+
+        public override Task CreateServerAsync(Func<GenericLoopbackServer, Uri, Task> funcAsync)
+        {
+            return LoopbackServer.CreateServerAsync((server, uri) => funcAsync(server, uri));
+        }
+
+        public override bool IsHttp11 => true;
+        public override bool IsHttp2 => false;
+    }
+
+    public sealed class Http2LoopbackServerFactory : LoopbackServerFactory
+    {
+        public static readonly Http2LoopbackServerFactory Singleton = new Http2LoopbackServerFactory();
+
+        public override async Task CreateServerAsync(Func<GenericLoopbackServer, Uri, Task> funcAsync)
+        {
+            using (var server = Http2LoopbackServer.CreateServer())
+            {
+                await funcAsync(server, server.Address);
+            }
+        }
+
+        public override bool IsHttp11 => false;
+        public override bool IsHttp2 => true;
+    }
+
+    public abstract class GenericLoopbackServer : IDisposable
+    {
+        // Accept a new connection, process a single request and send the specified response, and gracefully close the connection.
+        public abstract Task<HttpRequestData> HandleRequestAsync(HttpStatusCode statusCode = HttpStatusCode.OK, IList<HttpHeaderData> headers = null, string content = null);
+
+        public abstract void Dispose();
+    }
+
+    public struct HttpHeaderData
+    {
+        public string Name { get; }
+        public string Value { get; }
+
+        public HttpHeaderData(string name, string value)
+        {
+            Name = name;
+            Value = value;
+        }
+    }
```

Comment:
Nit: Could this just use `KeyValuePair<string, string>`?

Reply:
I could, but it gets really wordy. That's really the only reason I didn't.